# Preliminary EDA - Data Quality Hell

This notebook covers the initial inspection and cleaning of the **Model Case** dataset (January 1st - 15th, 2026). 

**Objective:** Prepare the data for deeper Exploratory Data Analysis (EDA) and future transformations using a stable, reproducible snapshot.

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 100)

## 1. Load Data

We start with the merged **Model Case** dataset containing 39,844 records from 19 countries extracted for the Jan 1-15 period.

In [2]:
input_csv = Path("../data/interim/all_jobs_merged.csv")
df = pd.read_csv(input_csv)
print(f"Initial Shape: {df.shape}")
df.head()

Initial Shape: (39844, 9)


country_code  \
0           at   
1           at   
2           at   
3           at   
4           at   

                                                                                           description  \
0  This position is posted by Jobgether on behalf of a partner company. We are currently looking fo...   
1  Product Owner - Data & Analytics (f/m/d) * Greentube GmbH * Vienna * Presence / Mobile * Publish...   
2  Im AI & Data Team bei Accenture vereinen wir einzigartige Persönlichkeiten und zukunftsgerich t ...   
3   Kitzbühel | Hybrides Arbeiten | Modebranche |  Datenstrategie | Interesse an AI willkommen Du l...   
4   Kitzbühel | Hybrides Arbeiten | Modebranche |  Datenstrategie | Interesse an AI willkommen Du l...   

                                                                          title  \
0                                                         Data Analyst (Growth)   
1                                      Product Owner - Data & Analytics (f/m/d)   
2                                       AI & Data Science Analyst (all genders)   
3         Data Analyst & Engineer (m/w/d) Retail & Marketing, Umzug nach Tirol?   
4  Senior Data Analyst (m/w/d) Google Analytics 4, Looker Studio, Power BI, SQL   

           id                           company  \
0  5586892007                         Jobgether   
1  5585358198                    Greentube GmbH   
2  5582988482                         Accenture   
3  5582082437  ANGEHEUERT GmbH Personalberatung   
4  5582082438  ANGEHEUERT GmbH Personalberatung   

                                                                                                 adref  \
0  eyJhbGciOiJIUzI1NiJ9.eyJzIjoiUU9EQVBLXzA4QkdDUS1YTGJsY0g4USIsImkiOiI1NTg2ODkyMDA3In0.t3OPQgJ6S0E...   
1  eyJhbGciOiJIUzI1NiJ9.eyJzIjoiUU9EQVBLXzA4QkdDUS1YTGJsY0g4USIsImkiOiI1NTg1MzU4MTk4In0.GEz4ubleXJ8...   
2  eyJhbGciOiJIUzI1NiJ9.eyJzIjoiUU9EQVBLXzA4QkdDUS1YTGJsY0g4USIsImkiOiI1NTgyOTg4NDgyIn0.k-K8zDLzgah...   
3  eyJhbGciOiJIUzI1NiJ9.eyJzIjoiUU9EQVBLXzA4QkdDUS1YTGJsY0g4USIsImkiOiI1NTgyMDgyNDM3In0.HjKCxxMdYSY...   
4  eyJhbGciOiJIUzI1NiJ9.eyJzIjoiUU9EQVBLXzA4QkdDUS1YTGJsY0g4USIsImkiOiI1NTgyMDgyNDM4In0.RQecAXsFJVE...   

           location               created   search_term  
0        Österreich  2026-01-15T21:42:20Z  Data Analyst  
1  Wien, Österreich  2026-01-15T02:15:16Z  Data Analyst  
2        Österreich  2026-01-13T21:37:41Z  Data Analyst  
3  Wien, Österreich  2026-01-13T11:32:57Z  Data Analyst  
4        Österreich  2026-01-13T11:32:57Z  Data Analyst

## 2. Drop Unnecessary Columns

Columns `description` (too inconsistent) and `adref` (no analytical value) are dropped to simplify the analysis.

> **Note:** We use `errors='ignore'` so the cell can be re-run without errors if the columns were already removed.

In [3]:
cols_to_drop = ['description', 'adref']
df = df.drop(columns=cols_to_drop, errors='ignore')
df.head()

country_code  \
0           at   
1           at   
2           at   
3           at   
4           at   

                                                                          title  \
0                                                         Data Analyst (Growth)   
1                                      Product Owner - Data & Analytics (f/m/d)   
2                                       AI & Data Science Analyst (all genders)   
3         Data Analyst & Engineer (m/w/d) Retail & Marketing, Umzug nach Tirol?   
4  Senior Data Analyst (m/w/d) Google Analytics 4, Looker Studio, Power BI, SQL   

           id                           company          location  \
0  5586892007                         Jobgether        Österreich   
1  5585358198                    Greentube GmbH  Wien, Österreich   
2  5582988482                         Accenture        Österreich   
3  5582082437  ANGEHEUERT GmbH Personalberatung  Wien, Österreich   
4  5582082438  ANGEHEUERT GmbH Personalberatung        Österreich   

                created   search_term  
0  2026-01-15T21:42:20Z  Data Analyst  
1  2026-01-15T02:15:16Z  Data Analyst  
2  2026-01-13T21:37:41Z  Data Analyst  
3  2026-01-13T11:32:57Z  Data Analyst  
4  2026-01-13T11:32:57Z  Data Analyst

## 3. Date Conversion

Converting the `created` column to a standard `datetime` format. We use `errors='coerce'` to handle any malformed strings.

In [4]:
df['created'] = pd.to_datetime(df['created'], errors='coerce')
print(f"Missing dates after conversion: {df['created'].isnull().sum()}")
df.info()

Missing dates after conversion: 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39844 entries, 0 to 39843
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype              
---  ------        --------------  -----              
 0   country_code  39844 non-null  object             
 1   title         39844 non-null  object             
 2   id            39844 non-null  int64              
 3   company       38511 non-null  object             
 4   location      39844 non-null  object             
 5   created       39844 non-null  datetime64[ns, UTC]
 6   search_term   35361 non-null  object             
dtypes: datetime64[ns, UTC](1), int64(1), object(5)
memory usage: 2.1+ MB


**Findings:**
- **General Scope:** The dataset contains **39,844 jobs** across 19 countries for the Jan 1-15 period.
- **Technical Depth:** Captures specialized roles (Data Engineer, Scientist, Analyst, MLOps, Architect).
- **Quality Metrics:**
    - **Titles:** Perfect coverage (0 nulls).
    - **Companies:** 1,333 missing values (3.3%), appearing across multiple territories.
    - **Dates:** All dates fall within the target range (Jan 1st - 15th).

**Proposed Strategy:** Fill missing companies with "Unknown". We also have a `search_term` column for 35,361 records to enable role-based segment analysis.

In [5]:
null_counts = df.isnull().sum()
print("--- Null Counts per Column ---")
print(null_counts)

print("\n--- Sample of Rows with Null Company (First 10) ---")
if 'company' in df.columns:
    display(df[df['company'].isnull()].head(10))

--- Null Counts per Column ---
country_code       0
title              0
id                 0
company         1333
location           0
created            0
search_term     4483
dtype: int64

--- Sample of Rows with Null Company (First 10) ---


country_code                                  title          id company  \
1921           au  Raw Materials Data Analyst - Rowville  5571054238     NaN   
6594           br                           Data Analyst  5586963668     NaN   
6595           br        Sales Operations & Data Analyst  5586961494     NaN   
6596           br              Analista De Powerbi Pleno  5586961442     NaN   
6597           br                           Data Analyst  5586961320     NaN   
6598           br                           Data Analyst  5586919512     NaN   
6599           br                           Data Analyst  5586918259     NaN   
6717           br         It Service Operations Engineer  5586965606     NaN   
6718           br                          Data Engineer  5586965000     NaN   
6719           br                          Data Engineer  5586963650     NaN   

                             location                   created    search_term  
1921              Rowville, Knox Area 2026-01-07 17:18:20+00:00   Data Analyst  
6594                 Manaus, Amazonas 2026-01-15 18:16:36+00:00   Data Analyst  
6595   São Paulo, Estado de São Paulo 2026-01-15 18:16:22+00:00   Data Analyst  
6596       Brasilia, Distrito Federal 2026-01-15 18:16:21+00:00   Data Analyst  
6597  Porto Alegre, Rio Grande do Sul 2026-01-15 18:16:20+00:00   Data Analyst  
6598   São Paulo, Estado de São Paulo 2026-01-15 18:11:53+00:00   Data Analyst  
6599   São Paulo, Estado de São Paulo 2026-01-15 18:11:40+00:00   Data Analyst  
6717   São Paulo, Estado de São Paulo 2026-01-15 18:16:48+00:00  Data Engineer  
6718   São Paulo, Estado de São Paulo 2026-01-15 18:16:44+00:00  Data Engineer  
6719  Piracicaba, Estado de São Paulo 2026-01-15 18:16:36+00:00  Data Engineer

**Status Check:** The data is mostly clean regarding mandatory fields (Title, Date, Location). The next step is to address internal consistency and redundant records.

## 4. Multi-Role & Record Redundancy Analysis

Checking for exact row duplicates and unique job identifier matches.

In [6]:
exact_duplicates = df.duplicated().sum()
id_duplicates = df.duplicated(subset=['id']).sum()

print(f"Exact row duplicates: {exact_duplicates}")
print(f"Duplicate job IDs: {id_duplicates}")

if id_duplicates > 0:
    print("\n--- Sample of Rows with Duplicate IDs ---")
    # Show some examples of duplicated IDs to understand why they exist
    duplicate_ids = df[df.duplicated(subset=['id'])]['id'].head(3)
    display(df[df['id'].isin(duplicate_ids)].sort_values(by='id').head(10))

Exact row duplicates: 4
Duplicate job IDs: 11819

--- Sample of Rows with Duplicate IDs ---


country_code                                            title  \
15             at  Senior Scientific Data Engineer- Vienna Austria   
25             at  Senior Scientific Data Engineer- Vienna Austria   
68             at  Senior Scientific Data Engineer- Vienna Austria   
613            at  Senior Scientific Data Engineer- Vienna Austria   
764            at  Senior Scientific Data Engineer- Vienna Austria   
1310           at  Senior Scientific Data Engineer- Vienna Austria   
31             at         Product Owner - Data & Analytics (f/m/d)   
1              at         Product Owner - Data & Analytics (f/m/d)   
739            at         Product Owner - Data & Analytics (f/m/d)   
853            at         Product Owner - Data & Analytics (f/m/d)   

              id         company          location                   created  \
15    5570582894    TetraScience  Wien, Österreich 2026-01-06 22:25:57+00:00   
25    5570582894    TetraScience  Wien, Österreich 2026-01-06 22:25:57+00:00   
68    5570582894    TetraScience  Wien, Österreich 2026-01-06 22:25:57+00:00   
613   5570582894    TetraScience  Wien, Österreich 2026-01-06 22:25:57+00:00   
764   5570582894    TetraScience  Wien, Österreich 2026-01-06 22:25:57+00:00   
1310  5570582894    TetraScience  Wien, Österreich 2026-01-06 22:25:57+00:00   
31    5585358198  Greentube GmbH  Wien, Österreich 2026-01-15 02:15:16+00:00   
1     5585358198  Greentube GmbH  Wien, Österreich 2026-01-15 02:15:16+00:00   
739   5585358198  Greentube GmbH  Wien, Österreich 2026-01-15 02:15:16+00:00   
853   5585358198  Greentube GmbH  Wien, Österreich 2026-01-15 02:15:16+00:00   

         search_term  
15      Data Analyst  
25    Data Architect  
68     Data Engineer  
613             Data  
764   Data Scientist  
1310             NaN  
31     Data Engineer  
1       Data Analyst  
739   Data Scientist  
853              NaN

**Findings:** While there are no exact row duplicates, we found **11,819 duplicate job IDs**.

**Revised Strategy:** Instead of treating these as 'junk' duplicates, we recognize that the same job can match multiple search terms (e.g., 'Data Engineer' and 'Big Data'). 

To preserve the richness of the classification, we will **not** drop these records yet. This allows us to perform a more accurate 'Market Demand' analysis by role in the next notebook.

## 5. Initial Cleaning (Execution)

Applying the decisions made above.

In [7]:
# Note: We are KEEPING duplicate IDs to preserve multi-role classification
# as decided in the Data Quality strategy update.

# Remove null title (safety check)
df = df.dropna(subset=['title'])

# Fill null companies
df['company'] = df['company'].fillna('Unknown')

print(f"Final Shape after cleaning: {df.shape}")
print(f"Total Unique Jobs (by ID): {df['id'].nunique()}")
print(f"\nRemaining missing values:\n{df.isnull().sum()}")

Final Shape after cleaning: (39844, 7)
Total Unique Jobs (by ID): 28025

Remaining missing values:
country_code       0
title              0
id                 0
company            0
location           0
created            0
search_term     4483
dtype: int64


## Next Steps

1. **Job Content Analysis:** Perform a frequency analysis of job titles to identify common roles.
2. **Geographical Distribution:** Visualize job density across the different countries.
3. **Time Series Exploration:** Analyze daily job posting counts to identify trends in early January.
4. **Company Profiling:** identify the top recruiters in this dataset.